File: app/src/main/java/com/bopr/android/smailer/ui/RemoteControlFragment.java
package com.bopr.android.smailer.ui;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;

import androidx.preference.Preference;

import com.bopr.android.smailer.GoogleAuthorizationHelper;
import com.bopr.android.smailer.R;

import static com.bopr.android.smailer.Settings.PREF_REMOTE_CONTROL_ACCOUNT;
import static com.bopr.android.smailer.Settings.PREF_REMOTE_CONTROL_ENABLED;
import static com.bopr.android.smailer.Settings.PREF_REMOTE_CONTROL_FILTER_RECIPIENTS;
import static com.bopr.android.smailer.Settings.PREF_REMOTE_CONTROL_NOTIFICATIONS;
import static com.bopr.android.smailer.util.TextUtil.isNullOrEmpty;
import static com.google.api.services.gmail.GmailScopes.MAIL_GOOGLE_COM;

public class RemoteControlFragment extends BasePreferenceFragment {

    private GoogleAuthorizationHelper authorizator;
    private Preference accountPreference;
    private SettingsListener settingsListener;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        authorizator = new GoogleAuthorizationHelper(this, PREF_REMOTE_CONTROL_ACCOUNT, MAIL_GOOGLE_COM);
        settingsListener = settings.registerChangeListener(new SettingsListener());
    }

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        addPreferencesFromResource(R.xml.pref_remote);

        // TODO: 24.02.2019 add help icon for remote control
        accountPreference = requirePreference(PREF_REMOTE_CONTROL_ACCOUNT);
        accountPreference.setOnPreferenceClickListener(preference -> {
            authorizator.selectAccount();
            return true;
        });
    }

    @Override
    public void onStart() {
        super.onStart();
        updateAccountPreference();
        updatePreferences();
    }

    @Override
    public void onDestroy() {
        settings.unregisterChangeListener(settingsListener);
        super.onDestroy();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        authorizator.onActivityResult(requestCode, resultCode, data);
    }

    private void updateAccountPreference() {
        String value = settings.getString(PREF_REMOTE_CONTROL_ACCOUNT, "");
        if (isNullOrEmpty(value)) {
            updateSummary(accountPreference, getString(R.string.not_specified), SUMMARY_STYLE_ACCENTED);
        } else if (!authorizator.isAccountExists(value)) {
            updateSummary(accountPreference, value, SUMMARY_STYLE_UNDERWIVED);
        } else {
            updateSummary(accountPreference, value, SUMMARY_STYLE_DEFAULT);
        }
    }

    private void updatePreferences() {
        boolean enabled = settings.getBoolean(PREF_REMOTE_CONTROL_ENABLED, false);
        accountPreference.setEnabled(enabled);
        requirePreference(PREF_REMOTE_CONTROL_NOTIFICATIONS).setEnabled(enabled);
        requirePreference(PREF_REMOTE_CONTROL_FILTER_RECIPIENTS).setEnabled(enabled);
    }

    private class SettingsListener implements SharedPreferences.OnSharedPreferenceChangeListener {

        @Override
        public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
            switch (key) {
                case PREF_REMOTE_CONTROL_ACCOUNT:
                    updateAccountPreference();
                    break;
                case PREF_REMOTE_CONTROL_ENABLED:
                    updatePreferences();
                    break;
            }

        }
    }

}

File: app/src/main/kotlin/com/bopr/android/smailer/ui/RemoteControlFragment.kt
package com.bopr.android.smailer.ui

import android.os.Bundle
import androidx.preference.Preference
import com.bopr.android.smailer.AccountsHelper.Companion.accounts
import com.bopr.android.smailer.R
import com.bopr.android.smailer.Settings.Companion.PREF_REMOTE_CONTROL_ACCOUNT
import com.bopr.android.smailer.Settings.Companion.PREF_REMOTE_CONTROL_ENABLED
import com.bopr.android.smailer.Settings.Companion.PREF_REMOTE_CONTROL_FILTER_RECIPIENTS
import com.bopr.android.smailer.Settings.Companion.PREF_REMOTE_CONTROL_NOTIFICATIONS
import com.bopr.android.smailer.Settings.Companion.settings
import com.bopr.android.smailer.control.mail.MailControlProcessor
import com.bopr.android.smailer.ui.InfoDialog.Companion.showInfoDialog
import com.bopr.android.smailer.util.PreferenceProgress
import com.bopr.android.smailer.util.SummaryStyle.SUMMARY_STYLE_ACCENTED
import com.bopr.android.smailer.util.SummaryStyle.SUMMARY_STYLE_UNDERWIVED
import com.bopr.android.smailer.util.getQuantityString
import com.bopr.android.smailer.util.requirePreference
import com.bopr.android.smailer.util.setOnChangeListener
import com.bopr.android.smailer.util.setOnClickListener
import com.bopr.android.smailer.util.showToast
import com.bopr.android.smailer.util.updateSummary
import com.google.api.services.gmail.GmailScopes.MAIL_GOOGLE_COM

// TODO: 24.02.2019 add help icon for remote control
/**
 * Remote control settings fragment.
 * 
 * @author Boris Pronin ([boris280471@gmail.com](mailto:boris280471@gmail.com))
 */
class RemoteControlFragment : BasePreferenceFragment(R.xml.pref_remote) {

    private lateinit var authorizationHelper: GoogleAuthorizationHelper

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        authorizationHelper = GoogleAuthorizationHelper(
            requireActivity(), PREF_REMOTE_CONTROL_ACCOUNT, MAIL_GOOGLE_COM
        )

        requirePreference(PREF_REMOTE_CONTROL_ACCOUNT).setOnClickListener {
            authorizationHelper.startAccountPicker()
        }

        requirePreference(PREF_PROCESS_SERVICE_MAIL).setOnClickListener {
            onProcessServiceMail(it)
        }

        requirePreference(PREF_REMOTE_CONTROL_ENABLED).setOnChangeListener {
            val value = settings.getBoolean(it.key)
            requirePreference(PREF_REMOTE_CONTROL_ACCOUNT).isEnabled = value
            requirePreference(PREF_REMOTE_CONTROL_NOTIFICATIONS).isEnabled = value
            requirePreference(PREF_REMOTE_CONTROL_FILTER_RECIPIENTS).isEnabled = value
            requirePreference(PREF_PROCESS_SERVICE_MAIL).isEnabled = value
        }

        requirePreference(PREF_REMOTE_CONTROL_ACCOUNT).setOnChangeListener {
            val account = settings.getString(it.key)
            it.apply {
                if (account.isNullOrEmpty()) {
                    updateSummary(R.string.unspecified, SUMMARY_STYLE_ACCENTED)
                } else if (!requireContext().accounts.isGoogleAccountExists(account)) {
                    updateSummary(account, SUMMARY_STYLE_UNDERWIVED)
                } else {
                    updateSummary(account)
                }
            }
        }
    }

    private fun onProcessServiceMail(preference: Preference) {
        val progress = PreferenceProgress(preference).apply { start() }
        MailControlProcessor(requireContext()).checkMailbox(
            onSuccess = { result ->
                progress.stop()
                showToast(
                    getQuantityString(R.plurals.mail_items, R.string.mail_items_zero, result)
                )
            },
            onError = { error ->
                progress.stop()
                showInfoDialog(
                    getString(R.string.remote_control),
                    error.message ?: error.toString()
                )
            }
        )
    }

    companion object {

        const val PREF_PROCESS_SERVICE_MAIL = "remote_control_process_service_mail"
    }
}